apps/rush-lib/src/logic/PackageJsonUpdater.ts

```diff
@@ -306,67 +309,98 @@ export class PackageJsonUpdater {
-        );
+      if(selectedVersion === undefined) {
+        console.log(`Querying registry for all versions of "${packageName}"...`);
 
-      let versionList: string[];
-      if (this._rushConfiguration.packageManager === 'yarn') {
-        versionList = JSON.parse(allVersions).data;
-      } else {
-        versionList = JSON.parse(allVersions);
-      }
+        let commandArgs: string[];
+        if (this._rushConfiguration.packageManager === 'yarn') {
+          commandArgs = ['info', packageName, 'versions', '--json'];
+        } else {
+          commandArgs = ['view', packageName, 'versions', '--json'];
+        }
 
-      console.log(colors.gray(`Found ${versionList.length} available versions.`));
+        const allVersions: string =
+          Utilities.executeCommandAndCaptureOutput(
+            this._rushConfiguration.packageManagerToolFilename,
+            commandArgs,
+            this._rushConfiguration.commonTempFolder
+          );
+
+        let versionList: string[];
+        if (this._rushConfiguration.packageManager === 'yarn') {
+          versionList = JSON.parse(allVersions).data;
+        } else {
+          versionList = JSON.parse(allVersions);
+        }
 
-      for (const version of versionList) {
-        if (semver.satisfies(version, initialSpec)) {
-          selectedVersion = initialSpec;
-          console.log(`Found a version that satisfies ${initialSpec}: ${colors.cyan(version)}`);
-          break;
+        console.log(colors.gray(`Found ${versionList.length} available versions.`));
+
+        for (const version of versionList) {
+          if (semver.satisfies(version, initialSpec)) {
+            selectedVersion = initialSpec;
+            console.log(`Found a version that satisfies ${initialSpec}: ${colors.cyan(version)}`);
+            break;
+          }
         }
-      }
 
-      if (!selectedVersion) {
-        throw new Error(`Unable to find a version of "${packageName}" that satisfies`
-          + ` the version specifier "${initialSpec}"`);
+        if (!selectedVersion) {
+          throw new Error(`Unable to find a version of "${packageName}" that satisfies`
+            + ` the version specifier "${initialSpec}"`);
+        }
       }
     } else {
       if (!this._rushConfiguration.ensureConsistentVersions) {
         console.log(colors.gray(`The "ensureConsistentVersions" policy is NOT active,`
           + ` so we will assign the latest version.`));
         console.log();
       }
-      console.log(`Querying NPM registry for latest version of "${packageName}"...`);
 
-      let commandArgs: string[];
-      if (this._rushConfiguration.packageManager === 'yarn') {
-        commandArgs = ['info', packageName, 'dist-tags.latest', '--silent'];
-      } else {
-        commandArgs = ['view', `${packageName}@latest`, 'version'];
+      // determine if the package is a project in the local repository
+      for (const rushConfigProject of rushConfigProjects) {
```

Comment:
This seems to be duplicated above (very similar). Better to move to its own function